### ISSUE TITLE
TypeSpec C# emitter doesn't produce LRO method with return type of non-generic Azure.Operation

### ISSUE THREAD

**Issue by devlie** [Azure/autorest.csharp#3965]
We are converting our Swagger to TypeSpec, and we have 2 LRO APIs, ImportDevices and ImportUpdate. The newly generated code changes the return type from `Operation` to `Operation<BinaryData>` so compilation breaks as it detects breaking change.

- [Old Swagger](https://github.com/Azure/azure-rest-api-specs-pr/blob/0788c5a7d5375a141d66f2cb82266f7c8190550d/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/stable/2022-10-01/deviceupdate.json#L843)
- [TypeSpec](https://github.com/Azure/azure-rest-api-specs-pr/blob/0788c5a7d5375a141d66f2cb82266f7c8190550d/specification/deviceupdate/Microsoft.DeviceUpdate/routes.tsp#L280)
- [Generated Swagger](https://github.com/Azure/azure-rest-api-specs-pr/blob/0788c5a7d5375a141d66f2cb82266f7c8190550d/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/preview/2023-10-01-preview/deviceupdate.json#L776)

In generated client, I only see this which returns generic Operation:
 > public virtual Operation<BinaryData> ImportDevices(WaitUntil waitUntil, RequestContent content, RequestContext context = null)

dotnet pack would return error:
``` 
C:\Users\leolie\.nuget\packages\microsoft.dotnet.apicompat\5.0.0-beta.20467.1\build\Microsoft.DotNet.ApiCompat.targets(82,5): error : MembersMustExist : Member 'public Azure.Operation Azure.IoT.DeviceUpdate.DeviceManagementClient.ImportDevices(Azure.WaitUntil, Azure.Core.RequestContent, Azure.RequestContext)' does
not exist in the implementation but it does exist in the contract. [D:\repos\azure-sdk-for-net\sdk\deviceupdate\Azure.IoT.DeviceUpdate\src\Azure.IoT.DeviceUpdate.csproj::TargetFramework=netstandard2.0] 
```

There is no way for me to workaround this as I cannot just hand-implement the old method since the method name and signature are identical except the return type.

**Comment by m-nash** [comment]
@devlie Do the other languages produce the same signature as before?  It seems that all languages would hit this issue where `getLroMetadata` is not returning `void` as the final result and therefore we are not generator `void`.

**Comment by weidongxu-microsoft** [comment]
@devlie

Do you expect "importUpdate" and "importDevices" to have final result?

From your Swagger, "importUpdate" defines an "Update" schema in response https://github.com/Azure/azure-rest-api-specs/blob/main/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/stable/2022-10-01/deviceupdate.json#L130-L144
while "importDevices" didn't define schema in response https://github.com/Azure/azure-rest-api-specs/blob/main/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/stable/2022-10-01/deviceupdate.json#L874-L882

PS: Java generated `SyncPoller<BinaryData, BinaryData>` for both in DPG from Swagger.

**Comment by devlie** [comment]
Not sure what you mean by final result. From REST API point of view, import update API returns 202 with operation-location header which points to the operation resource. When import is done operation resource would contain a link to the imported update.

For import device operation, IIRC there is no link to imported devices because usually there are too many devices.

**Comment by weidongxu-microsoft** [comment]
For "importUpdate", I assume you mean this "resourceLocation" which points to the imported update.
https://github.com/Azure/azure-rest-api-specs-pr/blob/0788c5a7d5375a141d66f2cb82266f7c8190550d/specification/deviceupdate/Microsoft.DeviceUpdate/models-update.tsp#L253-L254

**Comment by devlie** [comment]
> For "importUpdate", I assume you mean this "resourceLocation" which points to the imported update?
> 
> https://github.com/Azure/azure-rest-api-specs-pr/blob/0788c5a7d5375a141d66f2cb82266f7c8190550d/specification/deviceupdate/Microsoft.DeviceUpdate/models-update.tsp#L253-L254

Correct.

**Comment by archerzz** [comment]
@devlie You probably need to update typespec definition for `import:update`. Right now the latest parsing logic cannot find the return type for it. I guess it has something to do how you define resource location.
![image](https://github.com/Azure/autorest.csharp/assets/4516/f19a8780-5b33-4610-bc24-b9c67e91833f)

@markcowl  FYI

**Comment by weidongxu-microsoft** [comment]
`@finalOperation(DeviceUpdate.getUpdate)` over `importUpdate` maybe a solution to say an additional request is required for getting final result for `importUpdate`. Need further investigation/discussion.

**Comment by markcowl** [comment]
> `@finalOperation(DeviceUpdate.getUpdate)` over `importUpdate` maybe a solution to say an additional request is required for getting final result for `importUpdate`. Need further investigation/discussion.

@weidongxu-microsoft   There are two ways this can be done:
- If the client is intended to follow the uri returned in `UpdateOperation.resourceLocation`, then that property should get an `@finalLocation` decorator, and the property itself should be of type `ResourceLocation<T>`.  This indicates that clients should follow the link returned in this property, and that the finalResult will be `T`.

Note that, I have added an extra option to the `@finalLocation` decorator in my upcoming hotfix that allows you to set the intended final result there, without updating the type of the property - this could be a concern because ResourceLocation<T> indicates a url, which is a breaking change from string.

- if the client is just intended to make an additional call to the `getUpdate` , then adding an `@finalOperation` is appropriate, however, I do not think most DPG clients have this capability.  Also, you will need to specify a mechanism for calliung the operation - basically, how the operation paramaters for getUpdate can be filled out from the client, base on the request and response values for the original operation.

**Comment by devlie** [comment]
> @devlie You probably need to update typespec definition for `import:update`. Right now the latest parsing logic cannot find the return type for it. I guess it has something to do how you define resource location. ![image](https://private-user-images.githubusercontent.com/4516/289486410-f19a8780-5b33-4610-bc24-b9c67e91833f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMzE4MDYsIm5iZiI6MTcwMjMzMTUwNiwicGF0aCI6Ii80NTE2LzI4OTQ4NjQxMC1mMTlhODc4MC01YjMzLTQ2MTAtYmMyNC1iOWM2N2U5MTgzM2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMjE1MTQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmNjYjMyNDYyZWQ5ZGJhMGViZDM1NzdiYTM3M2ZiZTI5NzJiZjFiMzc2NWVjYjgwYjExMDgzZGZhN2E5ZjY2NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FBaN0_tabIR0kyURcnC0uxbMgyrAfy2_J8TNUg-jRJ0)
> 
> @markcowl FYI

@archerzz  your screenshot was for **import device operation** - that one doesn't have `resourceLocation` by design.

**Comment by weidongxu-microsoft** [comment]
> Note that import of large update can take anywhere between 30 mins to >1 hr, so a non blocking import update call followed by get (final) operation call might not be a bad idea.

SDK usually had such feature built-in into the class.
E.g. Java's `SyncPoller` would allow either user poll by themselves (say just leave it there for an hour, then call `poller.getFinalResult()`)

**Comment by weidongxu-microsoft** [comment]
Thanks Mark, yes, this works
```ts
  @finalLocation
  resourceLocation?: ResourceLocation<Update>;
```

Service probably need the extra option mentioned. From what I remember, the runtime value of this property is not a absolute URL. It would be a relative URL without host.

**Comment by archerzz** [comment]
This should be fixed after upgrading to `TypeSpec 0.51.0`. Blocked by https://github.com/Azure/autorest.csharp/pull/4022

**Comment by archerzz** [comment]
@devlie PR #4022 has been merged, which should have fixed this issue. I tried on my local, and the `importUpdate` operation has the correct signature in the generated codes. Could you please help to verify? Thanks.
![image](https://github.com/Azure/autorest.csharp/assets/4516/b669216f-aa97-4cc7-a8c1-fdcf6d911728)
